### PR TITLE
(#14841) Give useful error for array or hash lookup failure

### PR DESCRIPTION
**Pull request by pcarlisle**
Don't just throw a NoMethodError when using hiera_array or hiera_hash when nil is returned.

### CODE REVIEW

**Comment by kelseyhightower** [comment]
This is ready to merge once tests have been added.

File: lib/puppet/parser/functions/hiera_hash.rb [hunk 1]
```diff
@@ -30,7 +30,7 @@ module Puppet::Parser::Functions
 
         answer = hiera.lookup(key, default, hiera_scope, override, :hash)
 
-        raise(Puppet::ParseError, "Could not find data item #{key} in any Hiera data file and no default supplied") if answer.empty?
+        raise(Puppet::ParseError, "Could not find data item #{key} in any Hiera data file and no default supplied") if not answer or answer.empty?
```
Comment (pcarlisle) [inline lib/puppet/parser/functions/hiera_hash.rb]: Is the empty? check correct here? It seems to me like if you get back an empty array that's a valid return value.

Comment (kelseyhightower) [inline lib/puppet/parser/functions/hiera_hash.rb]: I think we should allow a Hash to be returned here, empty or not.

Comment (pcarlisle) [inline lib/puppet/parser/functions/hiera_hash.rb]: agreed, changed all of these to just check nil explicitly

**Comment by pcarlisle** [comment]
I think this is good for the basic fix. There is some disagreement here in the office about whether we should just let nil be returned instead of raising errors but I'm happy for now.

**Comment by pcarlisle** [comment]
Looks like this is merged into 1.x. Do we want to cherry-pick it into RC?

**Comment by kelseyhightower** [comment]
I just pushed it to the RC branch

On Thu, Jun 7, 2012 at 5:31 PM, pcarlisle <
reply@reply.github.com

> wrote:
> 
> Looks like this is merged into 1.x. Do we want to cherry-pick it into RC?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/puppetlabs/hiera-puppet/pull/31#issuecomment-6188629

## 

Kelsey Hightower
Developer
Puppet Labs
(678) 4719501

**Pull request merged**